### ISSUE TITLE
Problem with unconditional HTML encoding

### ISSUE THREAD

**Issue by gitsno** [turquoiseowl/i18n#202]
In issue #200 we corrected the call to Decode at the end of the nugget localizer, and I see that yesterday @turquoiseowl added more specific checking of msgids so that the msgid and message are only Decoded if the raw text doesn't match.

There is still a problem though in that the unconditional Encode at the end of nugget localizing can encode messages that were not intended to be encoded. This is a problem for messages that contain HTML tags, because the < > in the tags get encoded and the HTML then renders as text in the browser. 

The way it is now - you have to make sure nuggets do not include HTML tags, we should investigate a couple of approaches to making i18n handle HTML tags in nuggets, I'm thinking either escaping them or using a flag to only Encode if it was necessary to Decode (this would be a simple fix but it will only work when the msgid does not have mixed escaped HTML and raw HTML). 

**I'm also wondering if doing a simple Decode in the parser even makes sense**, what we're trying to decode is the nugget markup, so just doing a Decode and later an Encode is very heavy handed, what we really want to do is very specifically handle escaped nugget markup and leave everything else alone - that being the case we should not Decode and Encode at all, rather perform a specific search and replace for the encoded/escaped characters identified by NuggetBeginToken, NuggetEndToken, NuggetDelimiterToken, NuggetCommentToken, NuggetParameterBeginToken, and NuggetParameterEndToken.

Another issue is that it becomes very important to keep the PO generator and the nugget parser completely in sync on how this is implemented, otherwise the msgids that are generated can be different to what gets compared in the nugget parser.

A specific example of the problem:
If you have a nugget like this 
    [[[My &lt;b&gt;test&lt;/b&gt;]]]
it is going to get encoded by i18n, so that this gets sent to the browser
    My &amp;lt;test&amp;gt;
and that is clearly incorrect, it then shows the HTML tags as text in the browser instead of showing the bolded word **test**. In that case the msgid and the translated messages need to be able to have the < and > characters in them and they must not get encoded. Similarly, if I actually put &amp;lt; or &amp;gt; or any special character entities like &amp;raquo; in my nugget then they should be left alone, they should not get decoded.

**Comment by tomasr78** [comment]
I have found this issue when updated to 2.1.3.0 version. All my HTML tags are encoded from now. Any suggestion how to solve this? Are you planing to release update in near future?

**Comment by turquoiseowl** [comment]
My understanding of the current situation then WRT HTML encoding is as follows:

Originally, there was no Decoding/Encoding. This worked fine until #105: ASP.NET WebForms was HtmlEncoding the response BEFORE i18n got to it. This meant that any nugget that contained HTML-encodable characters (specifically, in this case, non-ASCII chars) was being altered such that the nugget parser was failing to pick it up.

For example, a nugget hard-coded into a Razor script as [[[Tate&Lyle]]] was arriving at the nugget parser as [[[Tate&amp;Lyle]]] (because ASP.NET WebForms was altering in the pipeline).

The fix for this was to get the string ```Tate&amp;Lyle``` picked up by the regex in the response, and convert back to ```Tate&Lyle``` (HtmlDecode) before looking it up in the message repository. That then matched what the PostBuild Parser found in the Razor script.

All seemed to work fine until #200 when it was noticed that we were HtmlDecoding the message output from the pipeline nugget parser. This was seen as incorrect and changed to HtmlEncode.

Okay. My first thought is why are we doing anything with the message string on output from the pipeline nugget parser? Why not just output as is? The ASP.NET WebForms fix should remain good. @gitsno Did you need the HtmlEncode?

**Comment by gitsno** [comment]
I don't need the decode or the encode (in fact for now I have gone in and commented them both out in my own branch) - the reason I brought up the decode bug in #200 was that it was a bug and was a problem for us, although as it turned out encoding wasn't a complete solution either.

I came to the party after #105 and haven't seen that issue (although I'd say that trying to deal with that issue in the nugget parser isn't the way to go because you're trying to deal with a specific case of encoding/decoding earlier in the pipeline and the way I see it, i18n is really based on the premise that it must literally translate everything it sees in one consistent way - that issue probably needs to be dealt with by doing some custom work in the drop down control or simply not using the drop down control at all).

I thought when I looked at the nugget parser that the reason for the decode and subsequent encode was to allow for use of the nugget wrapper with the nugget, and that's what I was proposing needs to be dealt without using HtmlDecode and HtmlEncode, but it sounds like that isn't what it was for. I have been working on a production release (using i18n, which has just gone live) so I haven't had a chance to test further, but I'd be fine with removing the HtmlDecode and HtmlEncode altogether.

I also saw that there is an HtmlDecode in one of the GetText overloads in the HttpContextExtension - any comment on what the intent of that is/was?

**Comment by turquoiseowl** [comment]
@gitsno Okay, thanks that. I've taken the plunge and removed all Html encoding and decoding, except for a fallback facility when the lookup for a nugget fails. In that case a second attempt at the lookup is made with HtmlDecode(msgid). If that fails too then the raw msgid will be used as the message.

The same lookup mechanism is employed by the HttpContext.GetText extension method with an optional parameter added to disable that.

No matter what, we don't HtmlEncode the message.

Should this approach do the trick, I'll push to NuGet in the next couple of days.

**Comment by tomasr78** [comment]
Could we expect an NuGet update this week? Thank you!

**Comment by turquoiseowl** [comment]
Version 2.1.4.0 now on NuGet. Hopefully everyone is happy...

**Comment by couraud** [comment]
Just discovered this after updating to v 2.1.4.0. 

In the past I used something like this in Razor: `@string.Format("[[[Your name is <b>%0</b>///{0}]]]", user.Name)`

Do I take it I now need to use Html.Raw(...) around this to avoid the HTML being encoded?

Thanks

**Comment by couraud** [comment]
Having dug around a bit more, I think I understand that my previous ability to simply use string.Format was an unintended benefit of #105. I guess the recommended approach is now to enclose all string.Formats with Html.Raw?

**Comment by turquoiseowl** [comment]
> I guess the recommended approach is now to enclose all string.Formats with Html.Raw?

Within the context of a Razor script and if the i18n message will possibly contain markup, then i guess so, yes.

**Comment by couraud** [comment]
Thanks for clarifying.
